### PR TITLE
test: use snapshot and split into multiple files

### PR DESCRIPTION
**Pull request by marcalexiei**
Part of #141

---

### 

### Split tests between files

Unit tests have been divided into multiple files and now use snapshots.
This should improve readability and, hopefully, maintenance efforts.
All common utilities and variables are now exported by `test/utils/index` file

### `insertStyle.test.ts` – replace `jsdom` with `happy-dom`

Since the test are quite simple I opted out to switch `jsdom` to `happy-dom` (less dependencies and faster)

## Additional changes

- All unit test dependencies have been updated:
  -  `ava`
  - `@ava/typescript`
  - `nyc`
  - `sinon`
- Added missing `@types` for
  -  `sinon`
  -  `icss-utils`

### CODE REVIEW

**Comment by elycruz** [comment]
Started looking over the changes - Will review them fully over the weekend.

**Comment by elycruz** [comment]
@marcalexiei Hey, the changes are pretty lenghty to go through - Could you summarize which tests have been updated to use snapshots, and which tests actually changed (I see a couple are using snapshots, but not all of them)?

**Review by marcalexiei — COMMENTED** [review]
The majority of test has been converted to snapshot, the change is basically replace one or more assertions with `t.snapshot`

---

Here is the list of list which are not using snapshot and why:
- `test/insertStyle.test.ts` do not use snapshot because is testing just the function itself
- `test/options.insert.test.ts` 
  - `should generate chunks with import insertStyle when insert is true` checks output. of bundle to ensure that insertStyle is used once (#139)
- `test/sourcemap.test.ts` doesn't use snapshots because are testing sourcemaps

---

After reviewing the changes I noticed tests in these files probably can use snapshots
- `test/options.output.test.ts`
- `test/watchList.test.ts` 

I'll update them now.

**Comment by marcalexiei** [comment]
## `test/watchList.test.ts`

Can't use snapshot since is comparing rollup properties with file system files. 
However I was able to simplify the logic (I created a separate commit).

## `test/options.output.test.ts`

Using snapshot here is not feasible due to rollup:
It looks like that, randomly, plugin `transform` is called first on `actual_b` and then on `actual_a` instead of calling transform on  `actual_a` and then on `actual_b`.

So instead of having:
```
body{color:red}body{color:green}
```
you end up having
```
body{color:green}body{color:red}
```
But only some times.

This looks like a rollup bug:
- https://github.com/rollup/rollup/issues/3888

Right now on `main` branch `should support output as function` sometimes fails for the exact same reason.
Also on other tests involving `output`, assertions are written using `indexOf !== -1` logic.
I assume it's because of this bug.

E.g.:
https://github.com/elycruz/rollup-plugin-sass/blob/0425739612e38ded0fda00e5060b25a735d620c3/test/index.test.ts#L265-L268

Right now there isn't much we can do. 
I'm going to add a code comment on `options.output.test.ts` referencing this PR

**Comment by elycruz** [comment]
I'm glad you noticed that bug (I had noticed it too) - Lets connect on the interwebs to discuss this PR - I'm on discord  (as edlcmoose) - **Note:** I'm only asking because this is a large change that we should probably do in smaller pull requests also the value of changing all tests to use snapshots isn't very clear to me so was hoping you could shed light on what the pros/cons are about adapting this approach, and/or if there are any analysis that you have performed on the approach.

**Comment by marcalexiei** [comment]
For  transparency and because I do not have a Discord account 😅 I will reply here:

 > I'm only asking because this is a large change that we should probably do in smaller pull requests

- I divided the monolithic  file into different files but the tests are the same.
- I haven't changed test description, so it can be compared with the previous implementation. 
- I had to move the variables that were on the top of `index.test.ts` inside a separate file (`test/utils/index.ts`) to avoid duplication between the new test files.
- The variables have been renamed with a `TEST_` prefix to avoid clash with src folder export.

I do not have any rush to get this merged so feel free to review when you have time 😀.

> [!NOTE] 
> If you stil find hard to review the changes I can close this PR and re-create smaller ones.

---

> the value of changing all tests to use snapshots isn't very clear

IMHO The main benefit is that though the snapshots you have a quick and fast way to compare the string output instead of having to search around for "expect" files that are read through `node:fs` and compared to plugin output after various trim / clean operations. 

---

>  if there are any analysis that you have performed on the approach

I used snapshot approach for the first time when implementing a custom webpack loader tests.
I find it more quick and easy than manipulating the strings directly via JS (which I was doing previously).

An additional benefit is that they can easily updated via `cli` command instead of performing the operation manually.

You can find the code here: https://github.com/marcalexiei/ractive-html-loader
The snapshot are used in this file: [test/loader.test.js](https://github.com/marcalexiei/ractive-html-loader/blob/main/test/loader.test.js)

---

**Comment by elycruz** [comment]
Ok, thank you for indulging me  -  I understand the changes better now - 
 will set time to review and validate them over the next couple of days.

**Comment by marcalexiei** [comment]
> thank you for indulging me

No problem 😉

> will set time to review and validate them over the next couple of days.

Thank you! As I said there is no rush 😃. 

---

Have a good weekend!

File: test/basic.test.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import test from "ava";
+import { rollup } from "rollup";
+import * as path from "path";
+
+import sass from "../src/index";
+import {
+  TEST_UTILS,
```
Comment (elycruz) [inline test/basic.test.ts:7]: So for TEST_UTILS, could you change the exports to just the methods themselves - Since they already live at 'test/utils' it becomes redundant to export the object `TEST_UTILS` - What would be good here would be to rename `squash` to `stripNewLines`, or something similar!

Comment (elycruz) [reply test/basic.test.ts:7]: Also, `getOutputFirstChunkCode` could just be `getFirstChunkCode` - since the declaration would read `getFirstChunkCode(output: ...) ...` - Would also change `output` to `chunks`, and/or, `outputChunks` for more correlation between what `output` constants actually represent.

**Review by elycruz — CHANGES_REQUESTED** [review]
Ok, so generally your changes look good, however they are to many to consume in one PR - You will need to separate it as follow, as well as make the review updates mentioned in the other review comments:

1.  [ ] One PR for dependency updates.
2.  [ ] One for changing all tests to use snapshots, and/or just contain updates, and,
3.  [ ] One for breaking the tests out into their own files.

Like this we would be able to consume the changes without requiring high cognitive load (too many things changing requires more things to keep in mind).

**Comment by elycruz** [comment]
@marcalexiei Also, the dependency updates can go together with the changing the tests PR.

**Comment by marcalexiei** [comment]
Ok...
